hazelcast/src/test/java/com/hazelcast/client/ClientConnectionProcessListenerTest.java

```diff
@@ -0,0 +1,221 @@
+                        .addAddress("nowhere.in.neverland:5701") // address not found
+                        .addAddress("localhost:6000") // port failure
+                )
+                .addListenerConfig(new ListenerConfig().setImplementation(listener));
+
+        Hazelcast.newHazelcastInstance(new Config().setClusterName(clusterName));
+
+        try {
+            HazelcastClient.newHazelcastClient(clientConfig);
+            fail("unexpectedly successful client startup");
+        } catch (IllegalStateException e) {
+            Address addr = new Address("localhost", 6000);
+            verify(listener, atLeastOnce()).hostNotFound("nowhere.in.neverland");
+            verify(listener, atLeastOnce()).possibleAddressesCollected(singletonList(addr));
+            verify(listener, atLeastOnce()).attemptingToConnectToAddress(addr);
+            verify(listener, atLeastOnce()).connectionAttemptFailed(addr);
+            verify(listener).clusterConnectionFailed(clusterName);
+            verifyNoMoreInteractions(listener);
+        }
+    }
+
+    @Test
+    public void remoteClosesConnection()
+            throws Exception {
+        ClientConnectionProcessListener listener = mock(ClientConnectionProcessListener.class);
+        ServerSocket server = new ServerSocket(5701);
+        try {
+            ForkJoinPool.commonPool().execute(() -> {
+                try {
+                    while (true) {
+                        Socket clientSocket = server.accept();
+                        Thread.sleep(500);
+                        OutputStream os = clientSocket.getOutputStream();
+                        os.write("junk response".getBytes(StandardCharsets.UTF_8));
+                        os.flush();
+                        os.close();
+                    }
+                } catch (IOException | InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+
+            ClientConfig clientConfig = newClientConfig()
+                    .setNetworkConfig(new ClientNetworkConfig().addAddress("localhost:5701"))
+                    .addListenerConfig(new ListenerConfig().setImplementation(listener));
+            HazelcastClient.newHazelcastClient(clientConfig);
+            fail("unexpectedly successful client startup");
+        } catch (IllegalStateException e) {
+            verify(listener, atLeastOnce()).possibleAddressesCollected(singletonList(new Address("localhost", 5701)));
+            verify(listener, atLeastOnce()).attemptingToConnectToAddress(new Address("localhost", 5701));
+            verify(listener, atLeastOnce()).remoteClosedConnection(new Address("localhost", 5701));
+            verify(listener).clusterConnectionFailed(getTestMethodName());
+            verifyNoMoreInteractions(listener);
+        } finally {
+            do {
+                server.close();
+            } while (!server.isClosed());
+        }
+    }
+
```

Comment:
there are still not tested apis left like `clientNotAllowedInCluster`

Reply:
added tests for `clientNotAllowedInCluster ` and `credentialsFailed`